File: sliding window/Longest Repeating Character Replacement.py
from collections import defaultdict
def characterReplacement(s: str, k: int) -> int:
    length = len(s)
    l, r = 0, 0
    seen = defaultdict(int)
    ans = 0
    frequent = 0
    while r < length:
        seen[s[r]] += 1
        frequent = max(frequent, seen[s[r]])

        while (r - l + 1 - frequent) > k:
            seen[s[l]] -= 1
            l += 1
        ans = max(ans, r - l + 1)
        r += 1

    return ans

print(characterReplacement("XYYX", 2))
print(characterReplacement("AAABABB", 1))

File: sliding window/Longest Repeating Character Replacement.cs
public class SolutionCharacterReplacement
{
    public int CharacterReplacement(string s, int k)
    {
        Dictionary<char, int> count = new Dictionary<char, int> { };

        int ans = 0;

        int start = 0, maxf = 0;

        for (int i = 0; i < s.Length; ++i)
        {
            if (count.ContainsKey(s[i]))
                ++count[s[i]];
            else
                count[s[i]] = 1;

            maxf = Math.Max(maxf, count[s[i]]);


            while ((i - start + 1) - maxf > k)
            {
                --count[s[start]];
                ++start;
            }
            ans = Math.Max(ans, i - start + 1);
        }

        return ans;
    }
}
